### PR TITLE
Added type hints - geoseries.py

### PR DESCRIPTION
**Pull request by steveberdy**
Added type hints.
See #1991

### CODE REVIEW

File: geopandas/geoseries.py
Comment (martinfleis) [inline geopandas/geoseries.py:866]:
```suggestion
        crs: Optional[Any] = None,
        epsg: Optional[int] = None,
        inplace: bool = False,
        allow_override: bool = False,
    ) -> GeoSeries:
```

File: geopandas/geoseries.py
Comment (martinfleis) [inline geopandas/geoseries.py:958]:
```suggestion
    def to_crs(self, crs: Optional[Any] = None, epsg: Optional[int] = None) -> GeoSeries:
```

File: geopandas/geoseries.py
Comment (martinfleis) [inline geopandas/geoseries.py:1078]:
```suggestion
    def to_json(self, **kwargs) -> str:
```
This returns JSON string.

**Review by martinfleis — CHANGES_REQUESTED** [review]
Hi @steveberdy,

Thanks for this. I have added some suggestions on how to handle different bits (untested). I think that we can go gradually, adding type hints to clear places and leaving others with more vague inputs for later additions.

I would suggest focusing on this PR and once this is approved and merged, moving to one of the other (say `geodataframe.py`). Do not try to update all at the same time.

File: geopandas/geoseries.py
Comment (steveberdy) [reply geopandas/geoseries.py:548]: Ok. My IDE automatically did that. My apologies.

File: geopandas/geoseries.py
Comment (martinfleis) [reply geopandas/geoseries.py]: @steveberdy You also need `from __future__ import annotations` suggested above to avoid the undefined error you're geting.

Comment (steveberdy) [reply geopandas/geoseries.py]: Alright.

Comment (steveberdy) [reply geopandas/geoseries.py]: @martinfleis I made those changes. I apologize for missing this

Comment (martinfleis) [reply geopandas/geoseries.py]: No problem, that is why we have a review process ;).

**Review by martinfleis — APPROVED** [review]
I think that this is a good start! We can always add some more type hints later. Thanks a lot, @steveberdy. Especially for the patience :).

If you want to move to another part, I'd suggest #2026.

**Comment by m-richards** [comment]
Hey @steveberdy I've just merged this back up to date with 0.11.1. Hopefully these PRs will get a bit more attention shortly for the 0.12 milestone.

Apologies, looks like I've main an error doing so on your fork, I'll fix that up.

**Comment by steveberdy** [comment]
@m-richards thanks for doing this!

File: geopandas/geoseries.py
Comment (m-richards) [inline geopandas/geoseries.py:529]:
```suggestion
    def __geo_interface__(self) -> Dict:
```
Using lowercase dict is a python 3.9 feature, so technically, we should use the one from typing for now
https://docs.python.org/3/whatsnew/3.9.html#type-hinting-generics-in-standard-collections

**Review by m-richards — APPROVED** [review]
I've had a look and I think I'm happy with all the additional annotations added (with one small suggestion).

I think at this point we should not let the perfect be the enemy of the good - all type annotations we add to GeoPandas are an improvement. As we continue with gradual typing we'll get more feedback from users about their correctness/ usability (we'll get indirect mypy feedback from users code, even if we aren't running mypy directly).

I think one of the nice obvious pieces of feedback we get is whether methods return something geospatial or not

**Comment by steveberdy** [comment]
Thank you @m-richards! I super appreciate this!

**Comment by martinfleis** [comment]
I gave it another look and can say that all these changes are correct. 

What I am less sure about is whether we want to merge them now and include in 0.13 or merge right after and try to include more type hints in 0.14. It is currently very patchy even within the GeoSeries context and I am not sure how I feel about that being part of a release.

**Comment by m-richards** [comment]
> I gave it another look and can say that all these changes are correct. 
> 
> What I am less sure about is whether we want to merge them now and include in 0.13 or merge right after and try to include more type hints in 0.14. It is currently very patchy even within the GeoSeries context and I am not sure how I feel about that being part of a release. 

When I merged it up to main I saw it was more patchy than I first realised. There are certainly some typing overloads we should add for things with inplace kwargs (but not necessarily as part of this PR). Given that, happy to defer to 0.14, as much as I think we should have type hints sooner rather than later.

**Comment by m-richards** [comment]
This was discussed on the community call today, agreed to move forward including this. Noted future things to deal with such as 
* typing overloads for e.g. inplace keyword
* arguments which take a GeoSeries / np.array of geometry / a list of geometries
will be deferred to future PRs

File: geopandas/geoseries.py
Comment (jorisvandenbossche) [inline geopandas/geoseries.py:39]: Just wondering because I am no typing expert, but how useful it is to annotate `crs` with `Optional[Any]` here? (because that basically gives no actual information. `Any` is not the default when untyped?)

Comment (m-richards) [reply geopandas/geoseries.py:39]: I can only really speak for mypy, semantically, mypy can track this union type separately:
```python
a : Optional[Any]
reveal_type(a) # note: Revealed type is "Union[Any, None]"
if a:
    reveal_type(a)   # note: Revealed type is "Any"
else:
    reveal_type(a) #  note: Revealed type is "Union[Any, None]"
```
But in practice I don't think this is useful since None is part of the Any type any way. So marking this optional won't help statically, it's only really a reminder to a programmer that it could be none. 

So there are probably two sensible options:
- replace this with purely Any
- Define a `CrsType = Union[int, pyproj.crs.CRS, Any]` which again is effectively just a reminder to a user and hope that in the future at some point pyproj has a more strict annotation on `from_user_input` 


Something else I was wondering which @jorisvandenbossche you might be able to weigh in on given pandas experience (if it has come up at all), should we use `from __future__ import annotations` or not. Originally I thought this would be "safe" since e.g. pandas has been using it for a long while, but I gather PEP 563 is still in limbo

Comment (martinfleis) [reply geopandas/geoseries.py:39]: Here's a [discussion](https://github.com/python/mypy/issues/3138) on this from my-yes repo, concluding that `Optional[Any]` is not the same as `Any`. I am personally fine keeping `Optional[Any]` in here, rather than specifying a subset like `Union[int, pyproj.crs.CRS, Any]` as it would have to be more something like `Union[int, str, dict, tuple pyproj.crs.CRS, Any]` at which point, `[Any]` is probably better.

> should we use from __future__ import annotations or not

afaik it is generally being used across the ecosystem (e.g. xarray).

**Comment by m-richards** [comment]
I've let this sit again, are there any objections to merging this?

**Comment by steveberdy** [comment]
Without looking at it, I'd say it looks great. Uhhh, hope that helps.

**Comment by martinfleis** [comment]
Let's go ahead and let's try to spend some time within the 0.14 milestone to get a better type coverage across the codebase.

**Comment by martinfleis** [comment]
Thanks @steveberdy and sorry it took so long to get the first part in!

**Comment by m-richards** [comment]
ah that's my  bad this is failing on main, should have re-merged master after the ruff PR was merged

**Comment by martinfleis** [comment]
@m-richards shall I open a PR fixing it? Or are you on it?

**Comment by m-richards** [comment]
If you wouldn't mind that would be great, I logged off for today, otherwise I can do tomorrow.

**Pull request merged**
